hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/historicalbalance/HistoricalBalanceService.java

```diff
@@ -0,0 +1,179 @@
+import com.hedera.mirror.importer.exception.ParserException;
+import com.hedera.mirror.importer.parser.record.RecordFileParsedEvent;
+import com.hedera.mirror.importer.parser.record.RecordFileParser;
+import com.hedera.mirror.importer.repository.AccountBalanceFileRepository;
+import com.hedera.mirror.importer.repository.AccountBalanceRepository;
+import com.hedera.mirror.importer.repository.RecordFileRepository;
+import com.hedera.mirror.importer.repository.TokenBalanceRepository;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Timer;
+import jakarta.inject.Named;
+import java.time.Instant;
+import java.util.concurrent.atomic.AtomicBoolean;
+import lombok.CustomLog;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.transaction.PlatformTransactionManager;
+import org.springframework.transaction.TransactionDefinition;
+import org.springframework.transaction.event.TransactionalEventListener;
+import org.springframework.transaction.support.TransactionTemplate;
+
+@ConditionalOnProperty(
+        prefix = "hedera.mirror.importer.parser.record.historical-balances",
+        name = "enabled",
+        matchIfMissing = true)
+@CustomLog
+@Named
+public class HistoricalBalanceService {
+
+    private final AccountBalanceFileRepository accountBalanceFileRepository;
+    private final AccountBalanceRepository accountBalanceRepository;
+    private final HistoricalBalancesProperties properties;
+    private final RecordFileRepository recordFileRepository;
+    private final AtomicBoolean running = new AtomicBoolean(false);
+    private final TokenBalanceRepository tokenBalanceRepository;
+    private final TransactionTemplate transactionTemplate;
+
+    // metrics
+    private final Timer generateDurationMetricFailure;
+    private final Timer generateDurationMetricSuccess;
+
+    public HistoricalBalanceService(
+            AccountBalanceFileRepository accountBalanceFileRepository,
+            AccountBalanceRepository accountBalanceRepository,
+            MeterRegistry meterRegistry,
+            PlatformTransactionManager platformTransactionManager,
+            HistoricalBalancesProperties properties,
+            RecordFileRepository recordFileRepository,
+            TokenBalanceRepository tokenBalanceRepository) {
+        this.accountBalanceFileRepository = accountBalanceFileRepository;
+        this.accountBalanceRepository = accountBalanceRepository;
+        this.properties = properties;
+        this.recordFileRepository = recordFileRepository;
+        this.tokenBalanceRepository = tokenBalanceRepository;
+
+        // Set repeatable read isolation level and transaction timeout
+        this.transactionTemplate = new TransactionTemplate(platformTransactionManager);
+        this.transactionTemplate.setIsolationLevel(TransactionDefinition.ISOLATION_REPEATABLE_READ);
```

Comment:
`@TransactionEventListener` states that `Interactions with the underlying transactional resource will not be committed in this phase`. So this transaction might be able to see some of the committed resources. Just to be safe:

```suggestion
        this.transactionTemplate.setIsolationLevel(TransactionDefinition.ISOLATION_REPEATABLE_READ);
        this.transactionTemplate.setPropagationBehavior(PROPAGATION_REQUIRES_NEW);
```